src/node/routes/vscode.ts

```diff
@@ -1,65 +1,105 @@
+import { logger } from "@coder/logger"
 import * as express from "express"
-import { DefaultedArgs } from "../cli"
+import { WebsocketRequest } from "../../../typings/pluginapi"
+import { logError } from "../../common/util"
+import { isDevMode } from "../constants"
 import { ensureAuthenticated, authenticated, redirect } from "../http"
 import { loadAMDModule } from "../util"
-import { Router as WsRouter, WebsocketRouter } from "../wsRouter"
+import { Router as WsRouter } from "../wsRouter"
 import { errorHandler } from "./errors"
 
-export interface VSServerResult {
-  router: express.Router
-  wsRouter: WebsocketRouter
-  codeServerMain: CodeServerLib.IServerAPI
-}
-
-export const createVSServerRouter = async (args: DefaultedArgs): Promise<VSServerResult> => {
-  // See ../../../vendor/modules/code-oss-dev/src/vs/server/main.js.
-  const createVSServer = await loadAMDModule<CodeServerLib.CreateServer>(
-    "vs/server/remoteExtensionHostAgent",
-    "createServer",
-  )
+export class CodeServerRouteWrapper {
+  /** Assigned in `ensureCodeServerLoaded` */
+  private _codeServerMain!: CodeServerLib.IServerAPI
+  private _wsRouterWrapper = WsRouter()
+  public router = express.Router()
 
-  const codeServerMain = await createVSServer(null, {
-    connectionToken: "0000",
-    ...args,
-    // For some reason VS Code takes the port as a string.
-    port: typeof args.port !== "undefined" ? args.port.toString() : undefined,
-  })
+  public get wsRouter() {
+    return this._wsRouterWrapper.router
+  }
 
-  const router = express.Router()
-  const wsRouter = WsRouter()
+  //#region Route Handlers
 
-  router.get("/", async (req, res, next) => {
+  private $root: express.Handler = async (req, res, next) => {
     const isAuthenticated = await authenticated(req)
+
     if (!isAuthenticated) {
       return redirect(req, res, "login", {
         // req.baseUrl can be blank if already at the root.
         to: req.baseUrl && req.baseUrl !== "/" ? req.baseUrl : undefined,
       })
     }
+
     next()
-  })
+  }
 
-  router.all("*", ensureAuthenticated, (req, res, next) => {
-    req.on("error", (error: any) => {
+  private $proxyRequest: express.Handler = async (req, res, next) => {
+    // We allow certain errors to propagate so that other routers may handle requests
```

Comment:
👏

❓ why start the name with `$`? is that some sort of special pattern?

Reply:
It’s something of a convention I’ve seen help with readability, similarly to prefixing private methods with `_`. Prefixing a method like `$proxyRequest` helps indicate a unique purpose compared to a more lengthy `proxyRequestHandler`